main/modes/picross/mode_picross.c

```diff
@@ -0,0 +1,1167 @@
+        // p->save->bankNames[i][1] = "i";
+        // p->save->bankNames[i][2] = "c";
+        // p->save->bankNames[i][3] = "_";
+        // p->save->bankNames[i][4] = "b";
+        // p->save->bankNames[i][5] = "_";
+        //im way to tired to look up how to case from int to char.
+        //just realized these can basically be random as long as they dont interfere with anyone elses code.
+    }
+
+    //Setup level
+    picrossSetupPuzzle(cont);
+}
+
+void picrossSetupPuzzle(bool cont)
+{   
+    wsg_t *levelwsg = &p->selectedLevel->levelWSG;
+
+    setCompleteLevelFromWSG(levelwsg);
+
+    //one step closer to this working correctly!
+    p->puzzle->width = levelwsg->w;
+    p->puzzle->height = levelwsg->h;
+    //maxhints = max(width/2 -1,height/2 -1) i ... think?
+
+    //rows
+    for(int i = 0;i<p->puzzle->height;i++)
+    {
+        p->puzzle->rowHints[i] = newHintFromPuzzle(i,true,p->puzzle->completeLevel);  
+    }
+    
+    //cols
+    for(int i = 0;i<p->puzzle->height;i++)
+    {
+        p->puzzle->colHints[i] = newHintFromPuzzle(i,false,p->puzzle->completeLevel);  
+    }
+
+    //Set the current level appropriately.
+    if(cont)
+    {
+        //Set the actual level from save data
+        loadPicrossProgress();
+    }else
+    {
+        //Set the actual level to be empty
+        for(int i = 0;i<p->puzzle->width;i++)
+        {
+            for(int j = 0;j<p->puzzle->height;j++)
+            {
+                p->puzzle->level[i][j] = SPACE_EMPTY;
+            }
+        }
+        savePicrossProgress();//clear out your previous (victory) data.
+        saveCompletedOnSelectedLevel(false);//clear out the fact that you have, perhaps, beaten this level before.
+    }
+
+    //Get chosen level and load it's data...
+
+    //Write the hint labels for each row and column
+
+    picrossResetInput();
+}
+
+//Scans the levelWSG and creates the finished version of the puzzle in the proper data format (2D array of enum).
+void setCompleteLevelFromWSG(wsg_t* puzz)
+{
+    // wsg_t puzz = puzzle;
+    //go row by row
+    paletteColor_t col = c555;
+    for(int r = 0;r<puzz->h;r++)//todo: puzzles that dont have the right size?
+    {
+        for(int c=0;c<puzz->w;c++){
+            col = puzz->px[(r * puzz->w) + c];
+            if(col == cTransparent || col == c555){
+                p->puzzle->completeLevel[c][r] = SPACE_EMPTY;
+            }else{
+                p->puzzle->completeLevel[c][r] = SPACE_FILLED;
+            }
+        }
+    }
+    
+}
+
+//lol i dont know how to do constructors, im a unity developer! anyway this constructs a source.
+//we use the same data type for rows and columns, hence the bool flag. index is the position in the row/col, and we pass in the level.
+//I guess that could be a pointer? Somebody more experienced with c, code review this and optimize please.
+picrossHint_t newHintFromPuzzle(uint8_t index, bool isRow, picrossSpaceType_t source[10][10])
+{
+    picrossHint_t t;
```

Comment:
I had to do some research on this, and I think it's OK.
https://stackoverflow.com/questions/9653072/return-a-struct-from-a-function-in-c
What I would have done, out of habit, is have this function return `picrossHint_t*`, `calloc()` a hint, and return a pointer to that.

I like the comment for the function, but I'd recommend writing it in the doxygen style to describe arguments. There are plugins to autogen those comment blocks

Reply:
I'm going to leave it as-is for now. Doxygen comments updated. Thanks for the input!